File: frontend/lib/mappers/reconciliation-mapper.ts
/**
 * Reconciliation Mapper - Stage 4 Reconciliation Intelligence Workspace
 *
 * Transforms backend DTOs to ViewModels for the Reconciliation Intelligence Workspace.
 * This is the ONLY location where DTO to ViewModel mapping occurs.
 *
 * Architecture Flow: Backend → API → DTO → Mapper → ViewModel
 */

import type {
  ReconciliationViewModel,
  DiscrepancyViewModel,
  StatusOverviewViewModel,
  AuditTrailEntryViewModel,
  ReconciliationSummaryViewModel,
  ReconciliationInsightViewModel,
  ReconciliationEvidenceChainViewModel,
  ReconciliationFiltersViewModel,
  ReconciliationNavigationViewModel,
} from '../../types/reconciliation-view-model';

// ===== DTO Types (from backend) =====

type ReconciliationStatus = 'pending' | 'confirmed' | 'rejected' | 'disputed';

interface DiscrepancyDTO {
  id: number;
  transaction_id: number;
  statement_id: number;
  type: string;
  expected_paise: number;
  actual_paise: number;
  difference_paise: number;
  status: ReconciliationStatus;
  notes?: string;
}

interface StatusOverviewDTO {
  total_transactions: number;
  reconciled: number;
  pending: number;
  discrepancies: number;
  match_rate: number;
}

interface AuditTrailEntryDTO {
  id: number;
  transaction_id: number;
  action: string;
  user: string;
  timestamp: string;
  notes?: string;
}

interface ReconciliationSummaryDTO {
  statement_id: number;
  bank: string;
  period_from: string;
  period_to: string;
  total_debit_paise: number;
  total_credit_paise: number;
  transaction_count: number;
  reconciled_count: number;
  status: ReconciliationStatus;
}

type ReconciliationInsightType = 'positive' | 'warning' | 'info' | 'alert';
type ReconciliationInsightSeverity = 'low' | 'medium' | 'high';

interface ReconciliationInsightDTO {
  type: ReconciliationInsightType;
  severity: ReconciliationInsightSeverity;
  message: string;
  action_url?: string;
}

interface ReconciliationEvidenceItemDTO {
  type: string;
  summary: string;
  source: string;
  confidence?: number;
}

interface ReconciliationCalculationStepDTO {
  name: string;
  description: string;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
}

interface ReconciliationEvidenceChainDTO {
  summary: string;
  evidence: ReconciliationEvidenceItemDTO[];
  calculation_steps: ReconciliationCalculationStepDTO[];
  source_references: string[];
  confidence_score: number;
}

interface ReconciliationDTO {
  statements: ReconciliationSummaryDTO[];
  discrepancies: DiscrepancyDTO[];
  status_overview: StatusOverviewDTO;
  audit_trail: AuditTrailEntryDTO[];
  insights: ReconciliationInsightDTO[];
  evidence_chain?: ReconciliationEvidenceChainDTO;
}

/**
 * Reconciliation Mapper Interface
 * Defines the contract for DTO to ViewModel mapping
 */
export interface IReconciliationMapper {
  /**
   * Map a single Reconciliation DTO to ViewModel
   */
  mapReconciliationDTO(dto: ReconciliationDTO): ReconciliationViewModel;

  /**
   * Map discrepancy DTOs to ViewModels
   */
  mapDiscrepancies(dtos: DiscrepancyDTO[]): DiscrepancyViewModel[];

  /**
   * Map status overview DTO to ViewModel
   */
  mapStatusOverview(dto: StatusOverviewDTO): StatusOverviewViewModel;

  /**
   * Map audit trail DTOs to ViewModels
   */
  mapAuditTrail(dtos: AuditTrailEntryDTO[]): AuditTrailEntryViewModel[];

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: ReconciliationInsightDTO[]): ReconciliationInsightViewModel[];

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: ReconciliationEvidenceChainDTO | null | undefined
  ): ReconciliationEvidenceChainViewModel | undefined;
}

/**
 * Reconciliation Mapper Implementation
 * Transforms backend reconciliation data to ViewModels
 */
export class ReconciliationMapper implements IReconciliationMapper {
  /**
   * Map a single Reconciliation DTO to ViewModel
   * @param dto - Reconciliation data from API
   * @returns ReconciliationViewModel for presentation
   */
  mapReconciliationDTO(dto: ReconciliationDTO): ReconciliationViewModel {
    return {
      statements: this.mapStatements(dto.statements),
      discrepancies: this.mapDiscrepancies(dto.discrepancies),
      status_overview: this.mapStatusOverview(dto.status_overview),
      audit_trail: this.mapAuditTrail(dto.audit_trail),
      insights: this.mapInsights(dto.insights),
      evidence_chain: this.mapEvidenceChain(dto.evidence_chain),
      filters: this.createDefaultFilters(),
      navigation: this.createDefaultNavigation(),
    };
  }

  /**
   * Map statement DTOs to ViewModels
   */
  private mapStatements(dtos: ReconciliationSummaryDTO[]): ReconciliationSummaryViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      statement_id: dto.statement_id,
      bank: dto.bank,
      period_from: dto.period_from,
      period_to: dto.period_to,
      total_debit_paise: dto.total_debit_paise,
      total_credit_paise: dto.total_credit_paise,
      transaction_count: dto.transaction_count,
      reconciled_count: dto.reconciled_count,
      status: dto.status,
    }));
  }

  /**
   * Map discrepancy DTOs to ViewModels
   */
  mapDiscrepancies(dtos: DiscrepancyDTO[]): DiscrepancyViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      id: dto.id,
      transaction_id: dto.transaction_id,
      statement_id: dto.statement_id,
      type: dto.type,
      expected_paise: dto.expected_paise,
      actual_paise: dto.actual_paise,
      difference_paise: dto.difference_paise,
      status: dto.status,
      notes: dto.notes,
    }));
  }

  /**
   * Map status overview DTO to ViewModel
   */
  mapStatusOverview(dto: StatusOverviewDTO): StatusOverviewViewModel {
    return {
      total_transactions: dto.total_transactions,
      reconciled: dto.reconciled,
      pending: dto.pending,
      discrepancies: dto.discrepancies,
      match_rate: dto.match_rate,
    };
  }

  /**
   * Map audit trail DTOs to ViewModels
   */
  mapAuditTrail(dtos: AuditTrailEntryDTO[]): AuditTrailEntryViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      id: dto.id,
      transaction_id: dto.transaction_id,
      action: dto.action,
      user: dto.user,
      timestamp: dto.timestamp,
      notes: dto.notes,
    }));
  }

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: ReconciliationInsightDTO[]): ReconciliationInsightViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      type: dto.type,
      severity: dto.severity,
      message: dto.message,
      action_url: dto.action_url,
    }));
  }

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: ReconciliationEvidenceChainDTO | null | undefined
  ): ReconciliationEvidenceChainViewModel | undefined {
    if (!dto) {
      return undefined;
    }
    return {
      summary: dto.summary,
      evidence: dto.evidence.map((item) => ({
        type: item.type,
        summary: item.summary,
        source: item.source,
        confidence: item.confidence,
      })),
      calculation_steps: dto.calculation_steps.map((step) => ({
        name: step.name,
        description: step.description,
        inputs: step.inputs,
        outputs: step.outputs,
      })),
      source_references: dto.source_references,
      confidence_score: dto.confidence_score,
    };
  }

  /**
   * Create default filters
   */
  private createDefaultFilters(): ReconciliationFiltersViewModel {
    return {
      status: undefined,
      banks: undefined,
      date_range: undefined,
    };
  }

  /**
   * Create default navigation
   */
  private createDefaultNavigation(): ReconciliationNavigationViewModel {
    return {
      deep_link: '/reconciliation',
      cross_references: {
        accounts: '/accounts',
        transactions: '/transactions',
      },
    };
  }
}

// Export singleton instance
export const reconciliationMapper = new ReconciliationMapper();

File: backend/src/core/mappers/reconciliation_mapper.py
"""Reconciliation Mapper

Transforms reconciliation domain objects into ReconciliationDTO instances.
This is the ONLY location where reconciliation API responses are constructed.
"""

from typing import Any

from src.core.dtos.reconciliation_dto import (
    AuditTrailEntryDTO,
    DiscrepancyDTO,
    ReconciliationDTO,
    ReconciliationInsightDTO,
    ReconciliationSummaryDTO,
    StatusOverviewDTO,
)


class ReconciliationMapper:
    """Mapper for reconciliation domain objects to DTOs."""

    @staticmethod
    def to_dto(reconciliation_data: dict[str, Any]) -> ReconciliationDTO:
        """Convert reconciliation data to ReconciliationDTO."""
        return ReconciliationDTO(
            statements=ReconciliationMapper._to_statement_summaries(
                reconciliation_data.get("statements", [])
            ),
            discrepancies=ReconciliationMapper._to_discrepancies(
                reconciliation_data.get("discrepancies", [])
            ),
            status_overview=ReconciliationMapper._to_status_overview_dto(
                reconciliation_data.get("status_overview", {})
            ),
            audit_trail=ReconciliationMapper._to_audit_trail(
                reconciliation_data.get("audit_trail", [])
            ),
            insights=ReconciliationMapper._to_insights(
                reconciliation_data.get("insights", [])
            ),
            evidence_chain=reconciliation_data.get("evidence_chain"),
        )

    @staticmethod
    def _to_statement_summaries(
        statements: list[dict[str, Any]],
    ) -> list[ReconciliationSummaryDTO]:
        """Convert statement data to ReconciliationSummaryDTO list."""
        return [
            ReconciliationSummaryDTO(
                statement_id=stmt.get("statement_id", 0),
                bank=stmt.get("bank", "Unknown"),
                period_from=stmt.get("period_from", ""),
                period_to=stmt.get("period_to", ""),
                total_debit_paise=stmt.get("total_debit_paise", 0),
                total_credit_paise=stmt.get("total_credit_paise", 0),
                transaction_count=stmt.get("transaction_count", 0),
                reconciled_count=stmt.get("reconciled_count", 0),
                status=stmt.get("status", "pending"),
            )
            for stmt in statements
        ]

    @staticmethod
    def _to_discrepancies(discrepancies: list[dict[str, Any]]) -> list[DiscrepancyDTO]:
        """Convert discrepancies to DiscrepancyDTO list."""
        return [
            DiscrepancyDTO(
                id=disc.get("id", 0),
                transaction_id=disc.get("transaction_id", 0),
                statement_id=disc.get("statement_id", 0),
                type=disc.get("type", "unknown"),
                expected_paise=disc.get("expected_paise", 0),
                actual_paise=disc.get("actual_paise", 0),
                difference_paise=disc.get("difference_paise", 0),
                status=disc.get("status", "pending"),
                notes=disc.get("notes"),
            )
            for disc in discrepancies
        ]

    @staticmethod
    def _to_status_overview_dto(overview_data: dict[str, Any]) -> StatusOverviewDTO:
        """Convert status overview data to StatusOverviewDTO."""
        return StatusOverviewDTO(
            total_transactions=overview_data.get("total_transactions", 0),
            reconciled=overview_data.get("reconciled", 0),
            pending=overview_data.get("pending", 0),
            discrepancies=overview_data.get("discrepancies", 0),
            match_rate=overview_data.get("match_rate", 0.0),
        )

    @staticmethod
    def _to_audit_trail(entries: list[dict[str, Any]]) -> list[AuditTrailEntryDTO]:
        """Convert audit trail entries to AuditTrailEntryDTO list."""
        return [
            AuditTrailEntryDTO(
                id=entry.get("id", 0),
                transaction_id=entry.get("transaction_id", 0),
                action=entry.get("action", ""),
                user=entry.get("user", "system"),
                timestamp=entry.get("timestamp", ""),
                notes=entry.get("notes"),
            )
            for entry in entries
        ]

    @staticmethod
    def _to_insights(insights: list[dict[str, Any]]) -> list[ReconciliationInsightDTO]:
        """Convert insights to ReconciliationInsightDTO list."""
        return [
            ReconciliationInsightDTO(
                type=insight.get("type", "info"),
                severity=insight.get("severity", "medium"),
                message=insight.get("message", ""),
                action_url=insight.get("action_url"),
            )
            for insight in insights
        ]
